python/lsst/afw/detection/multiband.py

```diff
@@ -0,0 +1,246 @@
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+__all__ = ["MultibandFootprint"]
+
+import numpy as np
+
+from lsst.geom import Point2I
+from lsst.afw.geom import SpanSet
+from lsst.afw.image import Mask, Image, MaskedImage, MultibandImage
+from lsst.afw.multiband import MultibandBase
+from . import Footprint, makeHeavyFootprint
+
+
+class MultibandFootprint(MultibandBase):
+    """Multiband Footprint class
+
+    A `MultibandFootprint` is a collection of HeavyFootprints that have
+    the same `SpanSet` and `peakCatalog` but different flux in each band.
+    """
+    def __init__(self, filters, singles=None, images=None, footprint=None, xy0=None, peaks=None, thresh=0):
+        """Initialize a `MultibandFootprint` object
+
+        Parameters
+        ----------
+        filters: list
+            List of filter names.
+            If `singles` is an `OrderedDict` then this argument is ignored,
+            otherwise it is required.
+        singles: list
+            A list of single band `HeavyFootprint` objects.
+            Either `singles` or `images` must be specified.
+            If `singles` is not `None`, then all arguments other than
+            `filters` are ignored.
+            Each `HeavyFootprint` should have the same `PeakCatalog` but
+            is allowed to have a different `SpanSet`, in which case the
+            `SpanSet` for each band is combined into the `SpanSet` of the
+            `MultibandFootprint`.
+        images: array, `MultibandImage`, or list of `afw.Image` objects
+            An array or `MultibandImage` (or collection of images in each band)
+            to convert into `HeavyFootprint` objects.
+            Only pixels above the `thresh` value for at least one band
+            will be included in the `SpanSet` and resulting footprints.
+        footprint: `Footprint`
+            `Footprint` that contains the `SpanSet` and `PeakCatalog`
+            to use for the `HeavyFootprint` in each band.
+        xy0: `Point2I`
+            If `images` is an array and `footprint` is `None` then specifying
+            `xy0` gives the location of the minimum `x` and `y` value of the
+            `images`.
+        peaks: `PeakCatalog`
+            Catalog containing information about the peaks located in the
+            footprints.
+        thresh: float or list of floats
+            Threshold in each band (or the same threshold to be used in all bands)
+            to include a pixel in the `SpanSet` of the `MultibandFootprint`.
+            If `Footprint` is not `None` then `thresh` is ignored.
+        """
+        if singles is None and images is not None:
+            from lsst.afw.image.utils import projectImage
```

Comment:
There's a circular dependence. `lsst.afw.utils` imports `lsst.afw.detection` for `clipImage`, so if the import is moved into the module scope it creates the error:
```
test_multiband.py:34: in <module>
    from lsst.afw.detection import GaussianPsf, Footprint, makeHeavyFootprint, MultibandFootprint, HeavyFootprintF
../python/lsst/afw/detection/__init__.py:26: in <module>
    from .multiband import *
../python/lsst/afw/detection/multiband.py:31: in <module>
    from lsst.afw.image.utils import projectImage
../python/lsst/afw/image/utils.py:29: in <module>
    import lsst.afw.detection as afwDetect
E   AttributeError: module 'lsst.afw' has no attribute 'detection'
```

It would perhaps be better to move the `lsst.afw.detection` import into the function in `lsst.afw.image.utils`, unless you have a clever way to avoid this problem.

Reply:
Given that `projectImage` is new (right?), how about moving it into `afw/multiband.py`?

Reply:
I feel like the ability to project a group of images into a common bounding box is useful outside of multiband objects and is really an image utility, not a multiband utility.

Reply:
Fair enough.  Unfortunately Python doesn't really provide a way to design for both dependency concerns and namespace concerns (i.e. the ability to "reopen" packages), so there isn't going to be a solution that really works well, and I'm happy to let you chose which evil is lesser.